client/state/data-layer/http-data.ts

```diff
@@ -0,0 +1,345 @@
+	) {
+		if ( 'development' === process.env.NODE_ENV && 'function' !== typeof dispatch ) {
+			throw new Error( 'Cannot use HTTP data without injecting Redux store enhancer!' );
+		}
+
+		const action: HttpDataAction = {
+			type: HTTP_DATA_REQUEST,
+			id: requestId,
+			fetch: 'function' === typeof fetchAction ? fetchAction() : fetchAction,
+			fromApi: 'function' === typeof fromApi ? fromApi : () => ( a: any ) => a,
+		};
+
+		dispatch ? dispatch( action ) : dispatchQueue.push( action );
+	}
+
+	return data;
+};
+
+interface Query {
+	[key: string]: Lazy< Resource >;
+}
+
+type Results< T extends Query > = { [P in keyof T]: ReturnType< T[P] > };
+
+/**
+ * Blocks execution until requested data has been fulfilled
+ *
+ *  - May return without data if data hasn't been fulfilled or failed
+ *  - Use for SSR contexts or when we _want_ to block rendering or execution
+ *    until the requested data has been fulfilled, e.g. when URL routing
+ *    depends on some data property
+ *  - _DO NOT USE_ when normal synchronous/data interactions suffice such
+ *    as is the case in 99.999% of React component contexts
+ *
+ * @example:
+ * waitForData( {
+ *     geo: () => requestGeoLocation(),
+ *     splines: () => requestSplines( siteId ),
+ * } ).then( ( { geo, splines } ) => {
+ *     return ( geo.state === 'success' || splines.state === 'success' )
+ *         ? res.send( renderToStaticMarkup( <LocalSplines geo={ geo.data } splines={ splines.data } /> ) )
+ *         : res.send( renderToStaticMarkup( <UnvailableData /> ) );
+ * }
+ *
+ * @param query - key/value pairs of data name and request
+ * @param timeout - how many ms to wait until giving up on requests
+ * @return fulfilled data of request (or partial if could not fulfill)
+ */
+export const waitForData = < T extends Query >(
+	query: T,
+	{ timeout }: { timeout?: number } = {}
+): Promise< Results< T > > =>
+	new Promise( ( resolve, reject ) => {
+		let unsubscribe = () => {};
+		let timer: TimerHandle;
+		const names = Object.keys( query );
+
+		const getValues = () =>
+			names.reduce(
+				( [ values, allBad, allDone ], name ) => {
+					const value = query[ name ]();
+
+					return [
+						{ ...values, [ name ]: value },
+						allBad && value.state === DataState.Failure,
+						allDone && ( value.state === DataState.Success || value.state === DataState.Failure ),
+					];
+				},
+				[ {}, true, true ]
+			);
+
+		const listener = () => {
+			const [ values, allBad, allDone ] = getValues();
+
+			if ( allDone ) {
+				clearTimeout( timer );
+				unsubscribe();
+				allBad ? reject( values ) : resolve( values as Results< T > );
+			}
+		};
+
+		if ( timeout ) {
+			timer = setTimeout( () => {
+				const [ values ] = getValues();
+
+				unsubscribe();
+				reject( values );
+			}, timeout );
+		}
+
+		unsubscribe = subscribe( listener );
+		listener();
+	} );
+
+if ( 'object' === typeof window && window.app && window.app.isDebug && window ) {
```

Comment:
The last `&& window` check doesn't seem to make any sense to me.

Reply:
Looks like a typo to me

```suggestion
if ( 'object' === typeof window && window.app && window.app.isDebug ) {
```